oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedStrategy.java

```diff
@@ -0,0 +1,424 @@
+        Preconditions.checkArgument(mongoDocBlockQueueSize > 0, "Property " + OAK_INDEXER_PIPELINED_MONGO_DOC_QUEUE_SIZE + " must be > 0. Was: " + mongoDocBlockQueueSize);
+        int mongoBatchSize = ConfigHelper.getSystemPropertyAsInt(OAK_INDEXER_PIPELINED_MONGO_DOC_BATCH_SIZE, DEFAULT_OAK_INDEXER_PIPELINED_MONGO_DOC_BATCH_SIZE);
+        Preconditions.checkArgument(mongoBatchSize > 0, "Property " + OAK_INDEXER_PIPELINED_MONGO_DOC_BATCH_SIZE + " must be > 0. Was: " + mongoBatchSize);
+        int transformThreads = ConfigHelper.getSystemPropertyAsInt(OAK_INDEXER_PIPELINED_TRANSFORM_THREADS, DEFAULT_OAK_INDEXER_PIPELINED_TRANSFORM_THREADS);
+        Preconditions.checkArgument(transformThreads > 0, "Property " + OAK_INDEXER_PIPELINED_TRANSFORM_THREADS + " must be > 0. Was: " + transformThreads);
+        int workingMemoryMB = ConfigHelper.getSystemPropertyAsInt(OAK_INDEXER_PIPELINED_WORKING_MEMORY_MB, DEFAULT_OAK_INDEXER_PIPELINED_WORKING_MEMORY_MB);
+        Preconditions.checkArgument(workingMemoryMB >= 0, "Property " + OAK_INDEXER_PIPELINED_WORKING_MEMORY_MB + " must be >= 0. Was: " + workingMemoryMB);
+        if (workingMemoryMB == 0) {
+            workingMemoryMB = autodetectWorkingMemoryMB();
+        }
+
+        int numberOfThreads = 1 + transformThreads + 1 + 1; // dump, transform, sort threads, sorted files merge
+        ExecutorService threadPool = Executors.newFixedThreadPool(numberOfThreads,
+                new ThreadFactoryBuilder().setNameFormat("mongo-dump").setDaemon(true).build()
+        );
+        // This executor can wait for several tasks at the same time. We use this below to wait at the same time for
+        // all the tasks, so that if one of them fails, we can abort the whole pipeline. Otherwise, if we wait on
+        // Future instances, we can only wait on one of them, so that if any of the others fail, we have no easy way
+        // to detect this failure.
+        ExecutorCompletionService ecs = new ExecutorCompletionService(threadPool);
+        ScheduledExecutorService monitorThreadPool = Executors.newScheduledThreadPool(1,
+                new ThreadFactoryBuilder().setNameFormat("monitor").setDaemon(true).build()
+        );
+        try {
+            int numberOfBuffers = 1 + transformThreads;
+            // The extra memory arena is to account for the memory taken by the SortKey entries that are part of the
+            // entry batches.
+            int memoryArenas = numberOfBuffers + 1;
+            int memoryForEntriesMB = workingMemoryMB / memoryArenas;
+            int maxNumberOfEntries = memoryForEntriesMB * 1024 * 4; // Assuming that 1KB is enough for 4 entries.
+            int maxNumberOfEntriesPerBuffer = maxNumberOfEntries / numberOfBuffers;
+
+            // A ByteBuffer can be at most Integer.MAX_VALUE bytes
+            int bufferSizeBytes;
```

Comment:
Just for readability purposes, could the assignment here with the local scope be refactored into a function?

Reply:
Done